hilti/include/rt/library.h

```diff
@@ -2,9 +2,46 @@
 
 #pragma once
 
+#include <string>
+
 #include <hilti/rt/result.h>
 #include <hilti/rt/util.h>
 
+namespace hilti::rt::library {
+
+/**
+ * Version information that's embedded as JSON into HLTO libraries to record
+ * the HILTI version they were compile with.
+ */
+struct Version {
+    std::string magic; /**< magic string for identification */
+    int hilti_version; /**< HILTI project version */
+    double created;    /**< time library was compiled in seconds since epoch */
+    bool debug;        /**< true if compiled in debug mode */
+    bool optimize;     /**< true if compiled with optimizations enabled */
+
+    std::filesystem::path path; /**< path to file that library was loaded from; not embedded into JSON, but filled in by
+                                   `Library::open()` */
+
+    /** Converts the instances into a JSON string. */
+    std::string toJSON() const;
+
+    /**
+     * Parsed a JSON representation of an instance.
+     *
+     * @return error if parsing of JSON failed
+     */
+    hilti::rt::Result<Nothing> fromJSON(const std::string& json);
+
+    /**
+     * Checks the version for compatiblity with the current runtime system.
+     * Prints out warnings on mismatches, but doesn't abort.
+     */
+    void checkCompatibility() const;
```

Comment:
This is hard to test for, could we instead return any errors? Something like
```suggestion
     * Returns an Error on mismatch, Nothing otherwise.
     */
    Result<Nothing> checkCompatibility() const;
```

Reply:
How would we return what's potentially multiple errors? Also, I'd prefer to keep having the  function report the warnings as that way reporting will be consistent across different callers.

Reply:
As discussed, leaving as is because it's hard to unit test this anyways.